### DISCUSSION TITLE
Events in Wave tables not working without pagination.

### DISCUSSION THREAD

**Discussion by senalw** [h2oai/wave#1993] [Q&A]
#### Wave SDK Version, OS

Wave SDK: 0.24.2 , OS: MacOs

#### Actual behavior

Events in `ui.table` not working without pagination.

https://user-images.githubusercontent.com/12801761/230857879-3bfb92ea-b19c-4e5b-b9ca-609a02ef505d.mov

#### Expected behavior

Any events that supported by `ui.table` should be worked even without pagination of tables.

#### Steps To Reproduce

You can run the following code to create the issue:

```
@app('/')
async def serve(q: Q):
    q.page['meta'] = ui.meta_card(box='')
    q.page['form'] = ui.form_card(box='1 1 4 4', items=[
        ui.table(name="sample",
                 columns=[
                     ui.table_column(
                         name="deployment_name",
                         label="Deployment Name",
                         sortable=True,
                         searchable=True,
                     ),
                     ui.table_column(
                         name="deployment_id",
                         label="Deployment ID",
                         sortable=True,
                     ),
                 ],
                 rows=[
                     ui.table_row(name="1",
                                  cells=["Deployment 1", "89u3489u98348393434"]),
                     ui.table_row(name="2",
                                  cells=["Deployment 2", "89u3489u98348394664"])
                 ],
                 pagination=ui.TablePagination(
                     total_rows=2,
                     rows_per_page=100,
                 ),
                 events=["sort"]
                 ),
    ])

    print(f"Events: {q.events}")

    await q.page.save()
```

**Comment by mturoci** [comment]
Hello @senalw. This is by design. The events should be fired only is `pagination` is set, otherwise all the actions shall be handled client-side by Wave itself.

**Comment by senalw** [comment]
Hi @mturoci, is this mentioned in the Wave documentation? 

I have a requirement of keeping time in `%Y/%m/%d %I:%M %p` format.  E.g. `2023/04/10 07:15 PM` 
To preserve above format, we have to use `string` data type instead of `time` data type for the column. How can we enable sorting for this column while pagination is disabled? (Lexicographical sorting doesn't work for this time format)

**Comment by mturoci** [comment]
> is this mentioned in the Wave documentation?

Added it here - https://github.com/h2oai/wave/commit/a1735a2dafe6cc2b9f8793e59217b8ba265ef2dd.

> I have a requirement of keeping time in %Y/%m/%d %I:%M %p format. E.g. 2023/04/10 07:15 PM

I would suggest pushing back on the requirement as it's not a good practice UX-wise. Wave table supports internationalization out of the box, meaning the date format is displayed based on the user's locale rather than the hardcoded format.

> How can we enable sorting for this column while pagination is disabled? (Lexicographical sorting doesn't work for this time format)

You can't. You can either use pagination or `time` type to have the sorting work properly.

**Comment by senalw** [comment]
@mturoci Thanks for the explanation.

> Wave table supports internationalization out of the box, meaning the date format is displayed based on the user's locale rather than the hardcoded format.

Ok. So this time format should be consistent in all over the application. For example, I have to show the time in a `ui.markup` or `ui.text` components and this time format should be consistent with the time format we show in the `ui.table`.  How can I do that?

**Comment by senalw** [comment]
@mturoci could you please assist on above?

**Comment by mturoci** [comment]
>  have to show the time in a ui.markup or ui.text components and this time format should be consistent with the time format we show in the ui.table

You need to use the same mechanism for displaying the date as used in table - internationalization (i18n).

**Comment by senalw** [comment]
> > have to show the time in a ui.markup or ui.text components and this time format should be consistent with the time format we show in the ui.table
> 
> You need to use the same mechanism for displaying the date as used in table - internationalization (i18n).

Yes, I'm asking do we have any mechanism to get the browser local time format from Wave SDK itself or do we have to depend on custom JS code?

**Comment by mturoci** [comment]
Try googling a bit before asking. That way, you can save us both a lot of time :) There are at least 2 ways of getting the user's locale:

* From the headers browser sends in the initial request.
* From JS.

Both are accessible in Wave.

**Comment by senalw** [comment]
> Try googling a bit before asking. That way, you can save us both a lot of time :) There are at least 2 ways of getting the user's locale:

Yes, I can find a way it get from JS by googling. But I just wanted to know whether is there a data type like `time` to use with `ui.text` and `ui.markup` components as we get in `ui.tables` ?
Even though Wave is open-source, I don't think we can google and find these kind of informations there. :slightly_smiling_face:

> * From the headers browser sends in the initial request.
> * From JS.

> Both are accessible in Wave.

Thanks for letting me know that both of above are accessible from Wave. 🙂

**Comment by senalw** [comment]
@shereezhang, as @mturoci mentioned in [this comment](https://github.com/h2oai/wave/issues/1916#issuecomment-1508001266), we can get the user's local browser time format using JS code and use it with `ui.markup` and `ui.text` components. But this is inconsistent compared to the mechanism  in `ui.table` of Wave SDK. Therefore I created #1936 feature request. 

CC @dulajra @ShehanIshanka
